dkulib/parallelizer/parallelizer.py

```diff
@@ -0,0 +1,254 @@
+        self, output_column_names: NamedTuple, row: Dict = None, batch: List[Dict] = None, **function_kwargs,
+    ) -> Union[Dict, List[Dict]]:  # sourcery skip: or-if-exp-identity
+        """Wrap a row-by-row or batch function with error logging and response parsing
+
+        It applies `self.function` and:
+        - If batch, parse the function response to extract results and errors using `self.batch_response_parser`
+        - handles errors from the function with two methods:
+            * (default) log the error message as a warning and return the row with error keys
+            * fail if there is an error (if `self.error_handling == ErrorHandling.FAIL`)
+
+        """
+        if row and batch:
+            raise (ValueError("Please use either row or batch as arguments, but not both"))
+        output = deepcopy(row) if row else deepcopy(batch)
+        for output_column in output_column_names:
+            if row:
+                output[output_column] = ""
+            else:
+                for output_row in output:
+                    output_row[output_column] = ""
+        try:
+            response = (
+                self.function(row=row, **function_kwargs) if row else self.function(batch=batch, **function_kwargs)
+            )
+            if row:
+                output[output_column_names.response] = response
+            else:
+                output = self.batch_response_parser(
+                    batch=batch, response=response, output_column_names=output_column_names
+                )
+                errors = [
+                    row[output_column_names.error_message] for row in output if row[output_column_names.error_message]
+                ]
+                if errors:
+                    raise BatchError(str(errors))
+        except self.exceptions_to_catch + (BatchError,) as error:
+            if self.error_handling == ErrorHandling.FAIL:
+                raise error
+            logging.warning(
+                f"Function {self.function.__name__} failed on: {row if row else batch} because of error: {error}"
+            )
+            error_type = str(type(error).__qualname__)
+            module = inspect.getmodule(error)
+            if module:
+                error_type = f"{module.__name__}.{error_type}"
+            if row:
+                output[output_column_names.error_message] = str(error)
+                output[output_column_names.error_type] = error_type
+                output[output_column_names.error_raw] = str(error.args)
+            else:
+                for output_row in output:
+                    output_row[output_column_names.error_message] = str(error)
+                    output_row[output_column_names.error_type] = error_type
+                    output_row[output_column_names.error_raw] = str(error.args)
+        return output
+
+    def _convert_results_to_df(
+        self, df: pd.DataFrame, results: List[Dict], output_column_names: NamedTuple,
```

Comment:
I did think about turning this into a class attribute, that is a good point. If we do that, it means we have to initialize the parallelizer with a dataframe (or a fixed list of columns). The consequence is that it would be harder for the user to re-use the same parallelizer instance on multiple dataframes. My assumption is that this scenario could happen. I could be wrong.

What do you think?

Reply:
So we store the column name values in a private class attribute?

Reply:
Yes sounds good

Reply:
Done in 475b191

Reply:
See caveats on https://github.com/dataiku/dss-plugin-dkulib/pull/10#discussion_r648406780